File: sdk/villow-0.1.0/villow-0.1.0/src/villow/agent.py
from __future__ import annotations

from collections.abc import Callable
from typing import Any
from urllib.parse import urljoin

import httpx

from villow.signing import sign_request


def task_template(slug: str):
    def decorator(func: Callable):
        setattr(func, "__villow_task_template__", slug)
        return func

    return decorator


def preview(slug: str):
    def decorator(func: Callable):
        setattr(func, "__villow_preview_template__", slug)
        return func

    return decorator


class Agent:
    def __init__(
        self,
        *,
        publisher_id: str = "publisher_local",
        agent_id: str = "agent_local",
        key_id: str = "key_local",
        secret: str = "local-secret",
        platform_base_url: str = "http://127.0.0.1:8000",
        transport: httpx.AsyncBaseTransport | httpx.BaseTransport | None = None,
    ) -> None:
        self.publisher_id = publisher_id
        self.agent_id = agent_id
        self.key_id = key_id
        self.secret = secret
        self.platform_base_url = platform_base_url.rstrip("/")
        self.transport = transport
        self.task_templates: dict[str, dict[str, str]] = {}
        self.prepare_handlers: dict[str, Callable[..., Any]] = {}
        self.run_handlers: dict[str, Callable[..., Any]] = {}
        self.preview_handlers: dict[str, Callable[..., Any]] = {}
        self._idempotency_store: dict[tuple[str, str], tuple[str, int, dict[str, Any]]] = {}
        self._answers_by_task: dict[str, dict[str, Any]] = {}
        self._delegated_by_task: dict[str, dict[str, bool]] = {}
        self._task_events: dict[str, list[dict[str, Any]]] = {}
        self._task_artifacts: dict[str, list[dict[str, Any]]] = {}
        self._task_states: dict[str, str] = {}
        # N1.5: last opaque continuation blob a handler returned, keyed by task (surfaced from /result).
        self._task_agent_state: dict[str, dict[str, Any]] = {}
        self._register_decorated_handlers()

    def context(self, **kwargs: Any):
        from villow.context import Context

        return Context(agent=self, **kwargs)

    def asgi_app(self):
        from villow.server import create_app

        return create_app(self)

    async def signed_platform_post(
        self,
        path_or_url: str,
        payload: dict[str, Any],
        *,
        idempotency_key: str,
    ) -> dict[str, Any]:
        url = self._absolute_url(path_or_url)
        body = _json_body(payload)
        headers = sign_request(
            method="POST",
            path=httpx.URL(url).path,
            body=body,
            key_id=self.key_id,
            secret=self.secret,
            publisher_id=self.publisher_id,
            agent_id=self.agent_id,
            idempotency_key=idempotency_key,
        )
        headers["content-type"] = "application/json"
        async with httpx.AsyncClient(transport=self.transport, timeout=30) as client:
            response = await client.post(url, content=body, headers=headers)
        response.raise_for_status()
        if not response.content:
            return {}
        return response.json()

    def _absolute_url(self, path_or_url: str) -> str:
        if path_or_url.startswith("http://") or path_or_url.startswith("https://"):
            return path_or_url
        return urljoin(f"{self.platform_base_url}/", path_or_url.lstrip("/"))

    def _register_decorated_handlers(self) -> None:
        for name in dir(self):
            member = getattr(self, name)
            if not callable(member):
                continue
            func = getattr(member, "__func__", member)
            slug = getattr(func, "__villow_task_template__", None)
            if slug:
                self.task_templates.setdefault(slug, {"slug": slug})
                if name == "prepare" or name.startswith("prepare_"):
                    self.prepare_handlers[slug] = member
                elif name in {"run", "execute"} or name.startswith("run_") or name.startswith("execute_"):
                    self.run_handlers[slug] = member
            preview_slug = getattr(func, "__villow_preview_template__", None)
            if preview_slug:
                self.task_templates.setdefault(preview_slug, {"slug": preview_slug})
                self.preview_handlers[preview_slug] = member


def _json_body(payload: dict[str, Any]) -> bytes:
    import json

    return json.dumps(payload, separators=(",", ":"), sort_keys=True).encode("utf-8")

File: sdk/villow-sdk-0.1.0/package/src/agent.ts
import { request } from 'undici';
import type { FastifyInstance } from 'fastify';

import { createApp } from './server.js';
import { signRequest } from './signing.js';

const TASK_TEMPLATE = Symbol('villow.taskTemplate');
const PREVIEW_TEMPLATE = Symbol('villow.previewTemplate');

type Handler = (...args: unknown[]) => unknown;
type DecoratedHandler = Handler & {
  [TASK_TEMPLATE]?: string;
  [PREVIEW_TEMPLATE]?: string;
};

export interface PlatformRequestInput {
  url: string;
  method: 'POST';
  headers: Record<string, string>;
  body: Buffer;
}

export interface PlatformResponse {
  statusCode: number;
  body: unknown;
}

export type PlatformRequest = (input: PlatformRequestInput) => Promise<PlatformResponse>;

export interface AgentOptions {
  publisherId?: string;
  agentId?: string;
  keyId?: string;
  secret?: string;
  platformBaseUrl?: string;
  platformRequest?: PlatformRequest;
}

export class Agent {
  publisherId: string;
  agentId: string;
  keyId: string;
  secret: string;
  platformBaseUrl: string;
  platformRequest?: PlatformRequest;
  taskTemplates = new Map<string, { slug: string }>();
  prepareHandlers = new Map<string, Handler>();
  runHandlers = new Map<string, Handler>();
  previewHandlers = new Map<string, Handler>();
  idempotencyStore = new Map<string, { bodyHash: string; statusCode: number; payload: Record<string, unknown> }>();
  answersByTask = new Map<string, Record<string, unknown>>();
  delegatedByTask = new Map<string, Record<string, boolean>>();
  taskEvents = new Map<string, Array<{ kind: string; payload: Record<string, unknown> }>>();
  taskArtifacts = new Map<string, Array<Record<string, unknown>>>();
  taskStates = new Map<string, string>();
  // N1.5: last opaque continuation blob a handler returned, keyed by task (surfaced from /result).
  taskAgentState = new Map<string, Record<string, unknown>>();

  constructor(options: AgentOptions = {}) {
    this.publisherId = options.publisherId ?? 'publisher_local';
    this.agentId = options.agentId ?? 'agent_local';
    this.keyId = options.keyId ?? 'key_local';
    this.secret = options.secret ?? 'local-secret';
    this.platformBaseUrl = (options.platformBaseUrl ?? 'http://127.0.0.1:8000').replace(/\/$/, '');
    this.platformRequest = options.platformRequest;
    this.registerDecoratedHandlers();
  }

  fastifyApp(): FastifyInstance {
    return createApp(this);
  }

  async signedPlatformPost(pathOrUrl: string, payload: Record<string, unknown>, options: { idempotencyKey: string }): Promise<Record<string, unknown>> {
    const url = this.absoluteUrl(pathOrUrl);
    const body = stableJsonBuffer(payload);
    const headers = signRequest({
      method: 'POST',
      path: new URL(url).pathname,
      body,
      keyId: this.keyId,
      secret: this.secret,
      publisherId: this.publisherId,
      agentId: this.agentId,
      idempotencyKey: options.idempotencyKey,
    });
    headers['content-type'] = 'application/json';

    const response = this.platformRequest
      ? await this.platformRequest({ url, method: 'POST', headers, body })
      : await defaultPlatformRequest({ url, method: 'POST', headers, body });

    if (response.statusCode < 200 || response.statusCode >= 300) {
      throw new Error(`platform_request_failed:${response.statusCode}`);
    }
    return (response.body ?? {}) as Record<string, unknown>;
  }

  private absoluteUrl(pathOrUrl: string): string {
    if (pathOrUrl.startsWith('http://') || pathOrUrl.startsWith('https://')) return pathOrUrl;
    return `${this.platformBaseUrl}/${pathOrUrl.replace(/^\//, '')}`;
  }

  private registerDecoratedHandlers(): void {
    const prototype = Object.getPrototypeOf(this) as Record<string, DecoratedHandler>;
    for (const name of Object.getOwnPropertyNames(prototype)) {
      if (name === 'constructor') continue;
      const original = prototype[name];
      if (typeof original !== 'function') continue;
      const taskSlug = original[TASK_TEMPLATE];
      if (taskSlug) {
        this.taskTemplates.set(taskSlug, { slug: taskSlug });
        const bound = original.bind(this);
        if (name === 'prepare' || name.startsWith('prepare')) {
          this.prepareHandlers.set(taskSlug, bound);
        } else if (['run', 'execute'].includes(name) || name.startsWith('run') || name.startsWith('execute')) {
          this.runHandlers.set(taskSlug, bound);
        }
      }
      const previewSlug = original[PREVIEW_TEMPLATE];
      if (previewSlug) {
        this.taskTemplates.set(previewSlug, { slug: previewSlug });
        this.previewHandlers.set(previewSlug, original.bind(this));
      }
    }
  }
}

export function taskTemplate(slug: string) {
  return function taskTemplateDecorator(
    valueOrTarget: unknown,
    _propertyKeyOrContext?: string | symbol | ClassMethodDecoratorContext,
    descriptor?: PropertyDescriptor,
  ): void | DecoratedHandler {
    if (typeof valueOrTarget === 'function' && descriptor === undefined) {
      (valueOrTarget as DecoratedHandler)[TASK_TEMPLATE] = slug;
      return valueOrTarget as DecoratedHandler;
    }
    if (descriptor?.value) {
      (descriptor.value as DecoratedHandler)[TASK_TEMPLATE] = slug;
    }
  };
}

export function preview(slug: string) {
  return function previewDecorator(
    valueOrTarget: unknown,
    _propertyKeyOrContext?: string | symbol | ClassMethodDecoratorContext,
    descriptor?: PropertyDescriptor,
  ): void | DecoratedHandler {
    if (typeof valueOrTarget === 'function' && descriptor === undefined) {
      (valueOrTarget as DecoratedHandler)[PREVIEW_TEMPLATE] = slug;
      return valueOrTarget as DecoratedHandler;
    }
    if (descriptor?.value) {
      (descriptor.value as DecoratedHandler)[PREVIEW_TEMPLATE] = slug;
    }
  };
}

async function defaultPlatformRequest(input: PlatformRequestInput): Promise<PlatformResponse> {
  const response = await request(input.url, { method: input.method, headers: input.headers, body: input.body });
  const text = await response.body.text();
  return { statusCode: response.statusCode, body: text ? (JSON.parse(text) as unknown) : {} };
}

export function stableJsonBuffer(payload: Record<string, unknown>): Buffer {
  return Buffer.from(JSON.stringify(sortJson(payload)));
}

function sortJson(value: unknown): unknown {
  if (Array.isArray(value)) return value.map(sortJson);
  if (value && typeof value === 'object') {
    return Object.fromEntries(Object.entries(value as Record<string, unknown>).sort(([a], [b]) => a.localeCompare(b)).map(([key, item]) => [key, sortJson(item)]));
  }
  return value;
}
